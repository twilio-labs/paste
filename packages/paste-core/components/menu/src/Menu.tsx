import * as React from 'react';
import {BoxElementProps, safelySpreadBoxProps, Box} from '@twilio-paste/box';
import {MenuPrimitive} from '@twilio-paste/menu-primitive';
import {MenuProps} from './types';

const StyledMenu = React.forwardRef<HTMLDivElement, BoxElementProps>(({style, ...props}, ref) => {
  return (
    <Box
      {...safelySpreadBoxProps(props)}
      backgroundColor="colorBackgroundBody"
      borderStyle="solid"
      borderWidth="borderWidth10"
      borderColor="colorBorderWeak"
      borderRadius="borderRadius20"
      boxShadow="shadow"
      maxWidth="size30"
      minWidth="size20"
      zIndex="zIndex20"
      paddingY="space30"
      _focus={{outline: 'none'}}
      style={style}
      ref={ref}
    />
  );
});

const Menu = React.forwardRef<HTMLDivElement, MenuProps>((props, ref) => {
  return <MenuPrimitive {...props} as={StyledMenu} ref={ref} />;
});
Menu.displayName = 'Menu';
export {Menu};
