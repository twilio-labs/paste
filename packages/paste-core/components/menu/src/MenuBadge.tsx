import { Badge } from "@twilio-paste/badge";
import type { BadgeProps } from "@twilio-paste/badge";
import { Box } from "@twilio-paste/box";
import { Button } from "@twilio-paste/button";
import type { ButtonProps } from "@twilio-paste/button";
import { ChevronDownIcon } from "@twilio-paste/icons/esm/ChevronDownIcon";
import { MenuPrimitiveButton } from "@twilio-paste/menu-primitive";
import type { MenuPrimitiveButtonProps } from "@twilio-paste/menu-primitive";
import { ScreenReaderOnly } from "@twilio-paste/screen-reader-only";
import * as React from "react";

export type MenuBadgeProps = MenuPrimitiveButtonProps &
  Omit<ButtonProps, "variant"> &
  Pick<BadgeProps, "variant"> & {
    i18nButtonLabel: string;
  };

export type MenuBadgeContentProps = Omit<MenuBadgeProps, "variant">;
export const MenuBadgeContent = React.forwardRef<HTMLButtonElement, MenuBadgeContentProps>(
  ({ children, element, i18nButtonLabel, ...props }, ref) => {
    return (
      <Box as="span" display="flex" flexDirection="row" columnGap="space20" alignItems="center">
        {children}
        <MenuPrimitiveButton
          {...props}
          variant="secondary_icon"
          size="reset"
          element={`${element}_BUTTON`}
          as={Button}
          ref={ref}
        >
          <ChevronDownIcon decorative={true} size="sizeIcon10" />
          <ScreenReaderOnly>{i18nButtonLabel}</ScreenReaderOnly>
        </MenuPrimitiveButton>
      </Box>
    );
  },
);
MenuBadgeContent.displayName = "MenuBadgeContent";

const MenuBadge = React.forwardRef<HTMLButtonElement, MenuBadgeProps>(
  ({ children, element = "MENU_BADGE", variant = "default", i18nButtonLabel, ...props }, ref) => {
    return (
      <Badge element={element} as="span" variant={variant}>
        <MenuBadgeContent {...props} element={element} i18nButtonLabel={i18nButtonLabel} ref={ref}>
          {children}
        </MenuBadgeContent>
      </Badge>
    );
  },
);

MenuBadge.displayName = "MenuBadge";

export { MenuBadge };
