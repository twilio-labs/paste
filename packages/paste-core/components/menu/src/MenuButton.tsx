import * as React from 'react';
import type {MenuPrimitiveButtonProps} from '@twilio-paste/menu-primitive';
import {MenuPrimitiveButton} from '@twilio-paste/menu-primitive';
import type {ButtonProps} from '@twilio-paste/button';
import {Button} from '@twilio-paste/button';

export type MenuButtonProps = MenuPrimitiveButtonProps & ButtonProps;

const MenuButton = React.forwardRef<HTMLButtonElement, MenuButtonProps>(({element = 'MENU_BUTTON', ...props}, ref) => {
  return (
    <MenuPrimitiveButton {...props} element={element} as={Button} ref={ref}>
      {props.children}
    </MenuPrimitiveButton>
  );
});
MenuButton.displayName = 'MenuButton';
export {MenuButton};
