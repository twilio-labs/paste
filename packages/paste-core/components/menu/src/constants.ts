export const MenuItemVariants = {
  DEFAULT: 'default',
  GROUP_ITEM: 'group_item',
} as const;
