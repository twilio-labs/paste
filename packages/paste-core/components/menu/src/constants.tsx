import * as React from 'react';
import {ProcessDisabledIcon} from '@twilio-paste/icons/esm/ProcessDisabledIcon';
import {ProcessDraftIcon} from '@twilio-paste/icons/esm/ProcessDraftIcon';
import {ProcessErrorIcon} from '@twilio-paste/icons/esm/ProcessErrorIcon';
import {ProcessInProgressIcon} from '@twilio-paste/icons/esm/ProcessInProgressIcon';
import {ProcessNeutralIcon} from '@twilio-paste/icons/esm/ProcessNeutralIcon';
import {ProcessSuccessIcon} from '@twilio-paste/icons/esm/ProcessSuccessIcon';
import {ProcessWarningIcon} from '@twilio-paste/icons/esm/ProcessWarningIcon';
import {ConnectivityAvailableIcon} from '@twilio-paste/icons/esm/ConnectivityAvailableIcon';
import {ConnectivityBusyIcon} from '@twilio-paste/icons/esm/ConnectivityBusyIcon';
import {ConnectivityUnavailableIcon} from '@twilio-paste/icons/esm/ConnectivityUnavailableIcon';
import {ConnectivityOfflineIcon} from '@twilio-paste/icons/esm/ConnectivityOfflineIcon';
import {ConnectivityNeutralIcon} from '@twilio-paste/icons/esm/ConnectivityNeutralIcon';

export const MenuItemVariants = {
  DEFAULT: 'default',
  DESTRUCTIVE: 'destructive',
  GROUP_ITEM: 'group_item',
  DESTRUCTIVE_GROUP_ITEM: 'destructive_group_item',
} as const;

export const StatusNames = {
  PROCESS_ERROR: 'ProcessError',
  PROCESS_WARNING: 'ProcessWarning',
  PROCESS_SUCCESS: 'ProcessSuccess',
  PROCESS_NEUTRAL: 'ProcessNeutral',
  PROCESS_IN_PROGRESS: 'ProcessInProgress',
  PROCESS_DISABLED: 'ProcessDisabled',
  PROCESS_DRAFT: 'ProcessDraft',
  AVAILABILITY_AVAILABLE: 'AvailabilityAvailable',
  AVAILABILITY_BUSY: 'AvailabilityBusy',
  AVAILABILITY_UNAVAILABLE: 'AvailabilityUnavailable',
  AVAILABILITY_NEUTRAL: 'AvailabilityNeutral',
  AVAILABILITY_OFFLINE: 'AvailabilityOffline',
} as const;

export const StatusIcons = {
  [StatusNames.PROCESS_NEUTRAL]: (
    <ProcessNeutralIcon color="colorTextIconNeutral" decorative={false} size="sizeIcon10" title="Neutral" />
  ),
  [StatusNames.PROCESS_IN_PROGRESS]: (
    <ProcessInProgressIcon color="colorTextIconNeutral" decorative={false} size="sizeIcon10" title="In progress" />
  ),
  [StatusNames.PROCESS_SUCCESS]: (
    <ProcessSuccessIcon color="colorTextIconSuccess" decorative={false} size="sizeIcon10" title="Success" />
  ),
  [StatusNames.PROCESS_WARNING]: (
    <ProcessWarningIcon color="colorTextIconWarning" decorative={false} size="sizeIcon10" title="Warning" />
  ),
  [StatusNames.PROCESS_ERROR]: (
    <ProcessErrorIcon color="colorTextIconError" decorative={false} size="sizeIcon10" title="Error" />
  ),
  [StatusNames.PROCESS_DISABLED]: (
    <ProcessDisabledIcon color="colorTextIcon" decorative={false} size="sizeIcon10" title="Disabled" />
  ),
  [StatusNames.PROCESS_DRAFT]: (
    <ProcessDraftIcon color="colorTextIcon" decorative={false} size="sizeIcon10" title="Draft" />
  ),
  [StatusNames.AVAILABILITY_AVAILABLE]: (
    <ConnectivityAvailableIcon color="colorTextIconAvailable" decorative={false} size="sizeIcon10" title="Available" />
  ),
  [StatusNames.AVAILABILITY_BUSY]: (
    <ConnectivityBusyIcon color="colorTextIconBusy" decorative={false} size="sizeIcon10" title="Busy" />
  ),
  [StatusNames.AVAILABILITY_UNAVAILABLE]: (
    <ConnectivityUnavailableIcon
      color="colorTextIconUnavailable"
      decorative={false}
      size="sizeIcon10"
      title="Unavailable"
    />
  ),
  [StatusNames.AVAILABILITY_OFFLINE]: (
    <ConnectivityOfflineIcon color="colorTextIconOffline" decorative={false} size="sizeIcon10" title="Offline" />
  ),
  [StatusNames.AVAILABILITY_NEUTRAL]: (
    <ConnectivityNeutralIcon color="colorTextIconNeutral" decorative={false} size="sizeIcon10" title="Neutral" />
  ),
};
