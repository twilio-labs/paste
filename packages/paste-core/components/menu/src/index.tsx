import * as React from 'react';
import {Box, BoxProps, safelySpreadBoxProps} from '@twilio-paste/box';
import {Text} from '@twilio-paste/text';
import {secureExternalLink} from '@twilio-paste/anchor';
import {Button, ButtonProps} from '@twilio-paste/button';
import {MediaObject, MediaFigure, MediaBody} from '@twilio-paste/media-object';
import {Separator, SeparatorProps} from '@twilio-paste/separator';
import {ChevronDownIcon} from '@twilio-paste/icons/esm/ChevronDownIcon';
import {
  useMenuPrimitiveState,
  MenuPrimitive,
  MenuPrimitiveProps,
  MenuPrimitiveItem,
  MenuPrimitiveItemProps,
  MenuPrimitiveButton,
  MenuPrimitiveButtonProps,
  MenuPrimitiveSeparator,
  MenuPrimitiveSeparatorProps,
} from '@twilio-paste/menu-primitive';

export type MenuProps = MenuPrimitiveProps & {'aria-label': string};
export interface MenuItemProps extends MenuPrimitiveItemProps {
  href?: string;
  as?: any;
}
export type MenuSeparatorProps = MenuPrimitiveSeparatorProps;
export type MenuButtonProps = MenuPrimitiveButtonProps & ButtonProps;
export type SubMenuButtonProps = MenuPrimitiveButtonProps;
interface StyledMenuItemProps extends BoxProps {
  href?: string;
}

const StyledMenu = React.forwardRef<HTMLDivElement, BoxProps>(({style, ...props}, ref) => {
  return (
    <Box
      {...safelySpreadBoxProps(props)}
      backgroundColor="colorBackgroundBody"
      borderStyle="solid"
      borderWidth="borderWidth10"
      borderColor="colorBorderLight"
      borderRadius="borderRadius20"
      boxShadow="shadowCard"
      maxWidth="size30"
      minWidth="size30"
      zIndex="zIndex10"
      _focus={{outline: 'none'}}
      style={style}
      ref={ref}
    />
  );
});

const StyledMenuItem = React.forwardRef<HTMLDivElement | HTMLAnchorElement, StyledMenuItemProps>(
  ({children, ...props}, ref) => {
    return (
      <Box
        {...(props.href && secureExternalLink(props.href))}
        {...safelySpreadBoxProps(props)}
        as="a"
        display="block"
        padding="space30"
        paddingLeft="space50"
        paddingRight="space50"
        textDecoration="none"
        _hover={{
          cursor: 'pointer',
        }}
        _focus={{
          outline: 'none',
          backgroundColor: 'colorBackgroundPrimaryLightest',
        }}
        _disabled={{cursor: 'not-allowed'}}
        ref={ref}
      >
        <Text
          as="div"
          color={props['aria-disabled'] ? 'colorTextWeaker' : 'colorTextLink'}
          textDecoration={props.tabIndex === 0 ? 'underline' : null}
        >
          {children}
        </Text>
      </Box>
    );
  }
);

const StyledMenuSeparator: React.FC<SeparatorProps> = props => {
  return <Separator {...props} orientation="horizontal" />;
};

const Menu = React.forwardRef<HTMLDivElement, MenuProps>((props, ref) => {
  return <MenuPrimitive {...props} as={StyledMenu} ref={ref} />;
});
Menu.displayName = 'Menu';
export {Menu};

const MenuItem = React.forwardRef<HTMLDivElement, MenuItemProps>(({as = StyledMenuItem, ...props}, ref) => {
  return <MenuPrimitiveItem {...props} as={as} ref={ref} />;
});
MenuItem.displayName = 'MenuItem';
export {MenuItem};

const MenuSeparator: React.FC<MenuSeparatorProps> = props => {
  // as prop from reakit for some reason only accepts a string of `hr` but accepts components. any prevent type errors
  return <MenuPrimitiveSeparator {...props} as={StyledMenuSeparator as any} />;
};
MenuSeparator.displayName = 'MenuSeparator';
export {MenuSeparator};

const MenuButton = React.forwardRef<HTMLButtonElement, MenuButtonProps>((props, ref) => {
  return (
    <MenuPrimitiveButton {...props} as={Button} ref={ref}>
      {props.children}
    </MenuPrimitiveButton>
  );
});
MenuButton.displayName = 'MenuButton';
export {MenuButton};

const SubMenuButton = React.forwardRef<HTMLButtonElement, SubMenuButtonProps>((props, ref) => {
  // MenuPrimitiveButton from reakit types `as` as HTML element names, but accepts components. any prevents type errors
  return (
    <MenuPrimitiveButton {...props} as={StyledMenuItem as any} ref={ref}>
      <MediaObject as="div" verticalAlign="center">
        <MediaBody as="div">{props.children}</MediaBody>
        <MediaFigure as="div" align="end" spacing="space20">
          <Box display="flex" transform="rotate(-90deg)">
            <ChevronDownIcon decorative size="sizeIcon20" />
          </Box>
        </MediaFigure>
      </MediaObject>
    </MenuPrimitiveButton>
  );
});
SubMenuButton.displayName = 'SubMenuButton';
export {SubMenuButton};

export {useMenuPrimitiveState as useMenuState};
