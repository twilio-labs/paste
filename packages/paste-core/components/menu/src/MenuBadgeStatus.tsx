import * as React from 'react';
import PropTypes from 'prop-types';
import {Box} from '@twilio-paste/box';
import type {ValueOf} from '@twilio-paste/types';

import {MenuBadge} from './MenuBadge';
import type {MenuBadgeProps} from './MenuBadge';
import {StatusIcons, StatusNames} from './constants';

export type MenuBadgeStatusVariants = ValueOf<typeof StatusNames>;

export interface MenuBadgeStatusProps extends Omit<MenuBadgeProps, 'variant'> {
  variant: MenuBadgeStatusVariants;
}

const getBadgeVariantFromStatus = (status: MenuBadgeStatusVariants): MenuBadgeProps['variant'] => {
  switch (status) {
    case StatusNames.PROCESS_NEUTRAL:
    case StatusNames.PROCESS_IN_PROGRESS:
      return 'neutral';
    case StatusNames.PROCESS_SUCCESS:
    case StatusNames.AVAILABILITY_AVAILABLE:
      return 'success';
    case StatusNames.PROCESS_WARNING:
    case StatusNames.AVAILABILITY_BUSY:
      return 'warning';
    case StatusNames.PROCESS_ERROR:
    case StatusNames.AVAILABILITY_UNAVAILABLE:
      return 'error';
    case StatusNames.PROCESS_DISABLED:
    case StatusNames.PROCESS_DRAFT:
    case StatusNames.AVAILABILITY_OFFLINE:
      return 'decorative10';
    case StatusNames.AVAILABILITY_NEUTRAL:
    default:
      return 'neutral';
  }
};

const MenuBadgeStatus = React.forwardRef<HTMLButtonElement, MenuBadgeStatusProps>(
  ({children, variant, ...props}, ref) => {
    const badgeVariant = React.useMemo(() => {
      return getBadgeVariantFromStatus(variant);
    }, [variant]);

    return (
      <MenuBadge {...props} variant={badgeVariant} ref={ref}>
        <Box display="flex" flexDirection="row" columnGap="space20" alignItems="center">
          {StatusIcons[variant]}
          {children}
        </Box>
      </MenuBadge>
    );
  }
);

MenuBadgeStatus.displayName = 'MenuBadgeStatus';

MenuBadgeStatus.propTypes = {
  children: PropTypes.node.isRequired,
};

export {MenuBadgeStatus};
