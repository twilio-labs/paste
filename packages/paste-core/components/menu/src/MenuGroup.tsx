import * as React from 'react';
import {safelySpreadBoxProps, Box} from '@twilio-paste/box';
import {Text} from '@twilio-paste/text';
import {MenuItemVariants} from './constants';
import type {MenuItemVariant, MenuGroupProps} from './types';

export const MenuGroupContext = React.createContext<MenuItemVariant>(MenuItemVariants.DEFAULT);

const MenuGroup = React.forwardRef<HTMLDivElement, MenuGroupProps>(({label, icon, children, ...props}, ref) => {
  return (
    <MenuGroupContext.Provider value={MenuItemVariants.GROUP_ITEM}>
      <Box {...safelySpreadBoxProps(props)} role="presentation" aria-label={label} textDecoration="none" ref={ref}>
        <Box display="flex" alignItems="center" paddingX="space70" paddingY="space30" cursor="default">
          {React.isValidElement(icon) ? (
            <Box flexShrink={0} size="sizeIcon30">
              {React.cloneElement(icon, {color: 'colorTextIcon'})}
            </Box>
          ) : null}
          <Text
            as="span"
            color="colorText"
            role="presentation"
            fontWeight="fontWeightBold"
            paddingLeft={icon != null ? 'space20' : undefined}
          >
            {label}
          </Text>
        </Box>
        {children}
      </Box>
    </MenuGroupContext.Provider>
  );
});
MenuGroup.displayName = 'MenuGroup';
export {MenuGroup};
