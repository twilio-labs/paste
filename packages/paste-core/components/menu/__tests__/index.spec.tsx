import * as React from 'react';
import {render, screen} from '@testing-library/react';
import {InformationIcon} from '@twilio-paste/icons/esm/InformationIcon';
import axe from '../../../../../.jest/axe-helper';
import {useMenuState, Menu, MenuItem, MenuButton, MenuButtonProps, MenuGroup, MenuSeparator} from '../src';

const PreferencesMenu = React.forwardRef<HTMLButtonElement, MenuButtonProps>((props, ref) => {
  const menu = useMenuState({baseId: 'sub-menu'});
  return (
    <>
      <MenuButton ref={ref} {...menu} {...props} data-testid="example-submenu-trigger">
        Preferences
      </MenuButton>
      <Menu {...menu} aria-label="Preferences" data-testid="example-submenu">
        <MenuItem {...menu} data-testid="example-submenu-item">
          Settings
        </MenuItem>
        <MenuItem {...menu} disabled data-testid="example-disabled-submenu-item">
          Extensions
        </MenuItem>
        <MenuSeparator {...menu} />
        <MenuItem {...menu}>Keyboard shortcuts</MenuItem>
      </Menu>
    </>
  );
});

const GROUP_LABEL_TEXT = 'Search Options';
const MenuMock: React.FC<{groupRef?: React.Ref<HTMLDivElement>}> = ({groupRef}) => {
  const menu = useMenuState({baseId: 'menu-example'});
  return (
    <>
      <MenuButton {...menu} variant="primary">
        Code
      </MenuButton>
      <Menu {...menu} aria-label="Code">
        <MenuItem {...menu} data-testid="example-menu-item">
          About Visual Studio Code
        </MenuItem>
        <MenuItem {...menu}>Check for Updates...</MenuItem>
        <MenuSeparator {...menu} data-testid="example-menu-separator" />
        <MenuItem {...menu} as={PreferencesMenu} />
        <MenuSeparator {...menu} data-testid="example-menu-separator-2" />
        <MenuGroup icon={<InformationIcon decorative />} label={GROUP_LABEL_TEXT} ref={groupRef}>
          <MenuItem {...menu}>Search with Google</MenuItem>
          <MenuItem {...menu} disabled>
            Search with Bing
          </MenuItem>
        </MenuGroup>
      </Menu>
    </>
  );
};

describe('Menu', () => {
  describe('Render', () => {
    it('should render', () => {
      const {asFragment} = render(<MenuMock />);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render a menu button with aria attributes', () => {
      render(<MenuMock />);
      const renderedMenuButton = screen.getByRole('button');
      expect(renderedMenuButton.getAttribute('aria-haspopup')).toEqual('menu');
      expect(renderedMenuButton.getAttribute('aria-controls')).toEqual('menu-example');
    });

    it('should render a menu', () => {
      render(<MenuMock />);
      const renderedMenu = screen.getByLabelText('Code');
      expect(renderedMenu.getAttribute('role')).toEqual('menu');
      expect(renderedMenu.getAttribute('aria-orientation')).toEqual('vertical');
    });

    it('should render a menu item', () => {
      render(<MenuMock />);
      const renderedMenuItem = screen.getByTestId('example-menu-item');
      expect(renderedMenuItem.getAttribute('role')).toEqual('menuitem');
    });

    it('should render a menu separator', () => {
      render(<MenuMock />);
      const renderedMenuItem = screen.getByTestId('example-menu-separator');
      expect(renderedMenuItem.getAttribute('aria-orientation')).toEqual('horizontal');
      expect(renderedMenuItem.tagName).toEqual('HR');
    });

    it('should render a menu group', () => {
      const groupRef = React.createRef<HTMLDivElement>();
      const {getByText} = render(<MenuMock groupRef={groupRef} />);

      expect(groupRef.current).not.toBeNull();
      // @ts-ignore we're checking it above
      expect(groupRef.current.getAttribute('aria-label')).toEqual(GROUP_LABEL_TEXT);

      const renderedGroupLabel = getByText(GROUP_LABEL_TEXT);
      expect(renderedGroupLabel.getAttribute('role')).toEqual('presentation');
    });

    it('should render a sub menu trigger', () => {
      render(<MenuMock />);
      const renderedSubMenuTrigger = screen.getByTestId('example-submenu-trigger');
      expect(renderedSubMenuTrigger.getAttribute('role')).toEqual('menuitem');
      expect(renderedSubMenuTrigger.getAttribute('aria-haspopup')).toEqual('menu');
      expect(renderedSubMenuTrigger.getAttribute('aria-expanded')).toEqual('false');
    });

    it('should render a submenu', () => {
      render(<MenuMock />);
      const renderedMenu = screen.getByTestId('example-submenu');
      expect(renderedMenu.getAttribute('role')).toEqual('menu');
      expect(renderedMenu.getAttribute('aria-orientation')).toEqual('vertical');
    });

    it('should render a sub menu item', () => {
      render(<MenuMock />);
      const renderedMenuItem = screen.getByTestId('example-submenu-item');
      expect(renderedMenuItem.getAttribute('role')).toEqual('menuitem');
    });

    it('should render a disabled sub menu item', () => {
      render(<MenuMock />);
      const renderedMenuItem = screen.getByTestId('example-disabled-submenu-item');
      expect(renderedMenuItem.getAttribute('role')).toEqual('menuitem');
      expect(renderedMenuItem.getAttribute('aria-disabled')).toEqual('true');
    });
  });

  describe('Accessibility', () => {
    it('Should have no accessibility violations', async () => {
      const {container} = render(<MenuMock />);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });
});
