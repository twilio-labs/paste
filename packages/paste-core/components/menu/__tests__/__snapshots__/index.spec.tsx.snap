// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu  Render should render 1`] = `
<DocumentFragment>
  .emotion-1 {
  position: relative;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: none;
  display: inline-block;
  outline: none;
  background: none;
  -webkit-transition: background-color 100ms ease-in,box-shadow 100ms ease-in;
  transition: background-color 100ms ease-in,box-shadow 100ms ease-in;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  box-shadow: inset 0 0 0;
  width: auto;
}

.emotion-1::-moz-focus-inner {
  border: none;
}

.emotion-1:hover,
.emotion-1:focus,
.emotion-1:active {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-1:hover {
  box-shadow: inset 0 0 0;
}

.emotion-1:focus {
  box-shadow: inset 0 0 0  ,;
}

.emotion-1:active {
  box-shadow: inset 0 0 0  ,;
}

.emotion-0 {
  display: grid;
  grid-auto-flow: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  vertical-align: middle;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  opacity: 100%;
}

<button
    aria-busy="false"
    aria-controls="menu-example"
    aria-expanded="false"
    aria-haspopup="menu"
    class="emotion-1"
    type="button"
  >
    <span
      class="emotion-0"
    >
      Code
    </span>
  </button>
  .emotion-6 {
  position: relative;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: none;
  display: inline-block;
  outline: none;
  background: none;
  -webkit-transition: background-color 100ms ease-in,box-shadow 100ms ease-in;
  transition: background-color 100ms ease-in,box-shadow 100ms ease-in;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  box-shadow: inset 0 0 0;
  width: auto;
}

.emotion-6::-moz-focus-inner {
  border: none;
}

.emotion-6:hover,
.emotion-6:focus,
.emotion-6:active {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-6:hover {
  box-shadow: inset 0 0 0;
}

.emotion-6:focus {
  box-shadow: inset 0 0 0  ,;
}

.emotion-6:active {
  box-shadow: inset 0 0 0  ,;
}

.emotion-5 {
  display: grid;
  grid-auto-flow: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  vertical-align: middle;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  opacity: 100%;
}

.emotion-14 {
  box-sizing: border-box;
  background-color: colorBackgroundBody;
  border-style: solid;
  border-width: borderWidth10;
  border-color: colorBorderLight;
  border-radius: borderRadius20;
  box-shadow: shadowCard;
  max-width: size30;
  min-width: size30;
  z-index: zIndex10;
}

.emotion-14:focus {
  outline: none;
}

.emotion-1 {
  box-sizing: border-box;
  display: block;
  padding: space30;
  padding-left: space50;
  padding-right: space50;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-1:hover {
  cursor: pointer;
}

.emotion-1:focus {
  outline: none;
  background-color: colorBackgroundPrimaryLightest;
}

.emotion-1:disabled,
.emotion-1:disabled:focus,
.emotion-1:disabled:hover,
.emotion-1[aria-disabled=true],
.emotion-1[aria-disabled=true]:focus,
.emotion-1[aria-disabled=true]:hover {
  cursor: not-allowed;
}

.emotion-0 {
  margin: 0;
  padding: 0;
  color: colorTextLink;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  font-size: fontSize30;
  line-height: lineHeight30;
}

.emotion-4 {
  box-sizing: border-box;
  margin: space0;
  border-width: borderWidth0;
  border-color: colorBorderLight;
  border-style: solid;
  border-bottom-width: borderWidth20;
  width: auto;
}

.emotion-9 {
  margin: 0;
  padding: 0;
  color: colorTextWeaker;
  font-size: fontSize30;
  line-height: lineHeight30;
}

.emotion-2 {
  margin: 0;
  padding: 0;
  color: colorTextLink;
  font-size: fontSize30;
  line-height: lineHeight30;
}

<div
    aria-label="Code"
    aria-orientation="vertical"
    class="emotion-14"
    data-dialog="true"
    hidden=""
    id="menu-example"
    role="menu"
    style="display: none;"
    tabindex="-1"
  >
    <a
      class="emotion-1"
      data-testid="example-menu-item"
      display="block"
      id="menu-example-1"
      role="menuitem"
      tabindex="0"
      text-decoration="none"
    >
      <div
        class="emotion-0"
        color="colorTextLink"
        font-size="fontSize30"
        text-decoration="underline"
      >
        About Visual Studio Code
      </div>
    </a>
    <a
      class="emotion-1"
      display="block"
      id="menu-example-2"
      role="menuitem"
      tabindex="-1"
      text-decoration="none"
    >
      <div
        class="emotion-2"
        color="colorTextLink"
        font-size="fontSize30"
      >
        Check for Updates...
      </div>
    </a>
    <hr
      aria-orientation="horizontal"
      class="emotion-4"
      data-testid="example-menu-separator"
      role="separator"
      width="auto"
    />
    <button
      aria-busy="false"
      aria-controls="sub-menu"
      aria-expanded="false"
      aria-haspopup="menu"
      class="emotion-6"
      data-testid="example-submenu-trigger"
      id="menu-example-3"
      role="menuitem"
      tabindex="-1"
      type="button"
    >
      <span
        class="emotion-5"
      >
        Preferences
      </span>
    </button>
    <div
      aria-label="Preferences"
      aria-orientation="vertical"
      class="emotion-14"
      data-dialog="true"
      data-testid="example-submenu"
      hidden=""
      id="sub-menu"
      role="menu"
      style="display: none;"
      tabindex="-1"
    >
      <a
        class="emotion-1"
        data-testid="example-submenu-item"
        display="block"
        id="sub-menu-1"
        role="menuitem"
        tabindex="0"
        text-decoration="none"
      >
        <div
          class="emotion-0"
          color="colorTextLink"
          font-size="fontSize30"
          text-decoration="underline"
        >
          Settings
        </div>
      </a>
      <a
        aria-disabled="true"
        class="emotion-1"
        data-testid="example-disabled-submenu-item"
        disabled=""
        display="block"
        id="sub-menu-2"
        role="menuitem"
        text-decoration="none"
      >
        <div
          class="emotion-9"
          color="colorTextWeaker"
          font-size="fontSize30"
        >
          Extensions
        </div>
      </a>
      <hr
        aria-orientation="horizontal"
        class="emotion-4"
        role="separator"
        width="auto"
      />
      <a
        class="emotion-1"
        display="block"
        id="sub-menu-3"
        role="menuitem"
        tabindex="-1"
        text-decoration="none"
      >
        <div
          class="emotion-2"
          color="colorTextLink"
          font-size="fontSize30"
        >
          Keyboard shortcuts
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
